src/ast/mod.rs

```diff
@@ -959,127 +802,157 @@ impl ToSexp for FunctionDecl {
+    Eq(Vec<GenericExpr<Head, Leaf, Ann>>),
+    Fact(GenericExpr<Head, Leaf, Ann>),
+}
+
+pub struct Facts<Head, Leaf, Ann>(pub Vec<GenericFact<Head, Leaf, Ann>>);
+
+#[allow(clippy::type_complexity)]
+impl<Head, Leaf, Ann> Facts<Head, Leaf, Ann>
+where
+    Head: Clone,
+    Leaf: Clone,
+    Ann: Clone,
+{
+    /// Flattens a list of facts into a Query.
+    /// For typechecking, we need the correspondence between the original ast
+    /// and the flattened one, so that we can annotate the original with types.
+    /// That's why this function produces a corresponding list of facts, annotated with
+    /// the variable names in the flattened Query.
+    /// (Typechecking preserves the original AST this way,
+    /// and allows terms and proof instrumentation to do the same).
+    pub(crate) fn to_query(
+        &self,
+        typeinfo: &TypeInfo,
+        fresh_gen: &mut impl FreshGen<Head, Leaf>,
+    ) -> (
+        Query<HeadOrEq<Head>, Leaf>,
+        Vec<GenericFact<(Head, Leaf), Leaf, Ann>>,
+    )
+    where
+        Leaf: SymbolLike,
+    {
+        let mut atoms = vec![];
+        let mut new_body = vec![];
+
+        for fact in self.0.iter() {
+            match fact {
+                GenericFact::Eq(exprs) => {
+                    let mut new_exprs = vec![];
+                    let mut to_equate = vec![];
+                    for expr in exprs {
+                        let (child_atoms, expr) = expr.to_query(typeinfo, fresh_gen);
+                        atoms.extend(child_atoms);
+                        to_equate.push(expr.get_corresponding_var_or_lit(typeinfo));
+                        new_exprs.push(expr);
+                    }
+                    atoms.push(GenericAtom {
+                        head: HeadOrEq::Eq,
+                        args: to_equate,
+                    });
+                    new_body.push(GenericFact::Eq(new_exprs));
+                }
+                GenericFact::Fact(expr) => {
+                    let (child_atoms, expr) = expr.to_query(typeinfo, fresh_gen);
+                    atoms.extend(child_atoms);
+                    new_body.push(GenericFact::Fact(expr));
+                }
             }
         }
+        (Query { atoms }, new_body)
     }
 }
 
-impl ToSexp for Fact {
+impl<Head: Display, Leaf: Display, Ann> ToSexp for GenericFact<Head, Leaf, Ann>
+where
+    Head: Display,
+    Leaf: Display,
+{
     fn to_sexp(&self) -> Sexp {
         match self {
-            Fact::Eq(exprs) => list!("=", ++ exprs),
-            Fact::Fact(expr) => expr.to_sexp(),
+            GenericFact::Eq(exprs) => list!("=", ++ exprs),
+            GenericFact::Fact(expr) => expr.to_sexp(),
         }
     }
 }
 
-impl Fact {
-    pub fn map_exprs(&self, f: &mut impl FnMut(&Expr) -> Expr) -> Fact {
+impl<Head, Leaf, Ann> GenericFact<Head, Leaf, Ann>
+where
+    Ann: Clone,
+    Head: Clone + Display,
+    Leaf: Clone + PartialEq + Eq + Display + Hash,
+{
+    pub(crate) fn map_exprs<Head2, Leaf2>(
```

Comment:
It does not. This was because some legacy implementations of `Expr` require these things (without a good reason), and all code built on top just followed this. I removed these unnecessary dependencies here and at some other places.